packages/react-compose/src/types.ts

```diff
@@ -13,23 +13,27 @@ export interface ShorthandConfig<TProps> {
 
 export type PropsOfElement<
   // tslint:disable-next-line:no-any
-  E extends keyof JSX.IntrinsicElements | React.JSXElementConstructor<any>
-> = JSX.LibraryManagedAttributes<E, React.ComponentPropsWithRef<E>>;
+  E extends keyof JSX.IntrinsicElements | React.JSXElementConstructor<any> | ComponentWithAs
+> = E extends { __PRIVATE_PROPS: any }
+  ? E['__PRIVATE_PROPS']
+  : JSX.LibraryManagedAttributes<E, React.ComponentPropsWithRef<E>>;
 
 // tslint:disable-next-line:interface-name
-export interface ComponentWithAs<TElementType extends React.ElementType = 'div', TProps = {}>
-  extends React.FunctionComponent {
-  <TExtendedElementType extends React.ElementType = TElementType>(
-    props: Omit<PropsOfElement<TExtendedElementType>, 'as' | keyof TProps> & { as?: TExtendedElementType } & TProps,
-  ): JSX.Element | null;
-  displayName?: string;
-
-  defaultProps?: Partial<TProps & { as: TElementType }>;
+export type ComponentWithAs<TElementType extends keyof JSX.IntrinsicElements = 'div', TProps = {}> = (<
```

Comment:
I reverted this thing to be case because of strange resolutions of TS: https://github.com/microsoft/TypeScript/issues/39313

***

`ComponentWithAs` is good enough to be used of other our components even without `compose()`, but before this change it marked `props` as `any`:

```tsx
const Divider: ComponentWithAs<'div', DividerProps> = (props /* becomes any 💣 */) => {}
```